otherlibs/dune-rpc/private/dune_rpc_private.ml

```diff
@@ -468,32 +487,49 @@ module Client = struct
       let next t =
         let* () = Fiber.return () in
         check_active t;
-        if t.next_pending then
+        (match t.pending_request_id with
+        | Some _ ->
           Code_error.raise "Poll.next: previous Poll.next did not terminate yet"
-            [];
-        t.next_pending <- true;
+            []
+        | None -> ());
         let id =
           Sexp.record
             [ ("poll", Id.to_sexp t.id)
             ; ("i", Sexp.Atom (string_of_int t.counter))
             ]
           |> Id.make
         in
+        t.pending_request_id <- Some id;
         let+ res = request ~id t.client t.poll t.id in
-        t.next_pending <- false;
+        t.pending_request_id <- None;
         match res with
-        | Ok res -> res
-        | Error e ->
+        | `Cancelled -> None
+        | `Ok (Ok res) -> res
+        | `Ok (Error e) ->
           (* cwong: Should this really be a raise? *)
           raise (Response.Error.E e)
 
       let cancel t =
         let* () = Fiber.return () in
         check_active t;
         t.active <- false;
-        notification t.client t.cancel t.id
+        (* XXX should we add a pool to stop waiting for the notification to
```

Comment:
We have an api to explicitly detach fibers. See `Fiber.Pool`. It's a bit of a pain to use it since it requires users of the library to provide an implementation to apply the functor.

Reply:
I guess one way to see the benefit of a pool would be to spam cancel requests?

Reply:
Not necessarily. Instead of waiting for the server to ack the cancellation, you can proceed. In any case, the caller can put the cancellation in his own pool without much difficulty.